gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientWriteChannel.java

```diff
@@ -196,31 +121,23 @@ public void close() throws IOException {
       writableByteChannel.close();
     } catch (Exception e) {
       GoogleCloudStorageEventBus.postOnException();
-      throw new IOException(String.format("Upload failed for '%s'", resourceId), e);
+      throw new IOException(
+          String.format("Upload failed for '%s'. reason=%s", resourceId, e.getMessage()), e);
     } finally {
       writableByteChannel = null;
     }
   }
 
-  @Override
-  public void handleResponse(Boolean response) {
-    this.uploadSucceeded = response;
-  }
-
-  @Override
-  protected String getResourceString() {
-    return resourceId.toString();
-  }
-
-  public boolean isUploadSuccessful() {
-    return uploadSucceeded;
-  }
-
   private int writeInternal(ByteBuffer byteBuffer) throws IOException {
     int bytesWritten = writableByteChannel.write(byteBuffer);
     logger.atFinest().log(
         "%d bytes were written out of provided buffer of capacity %d",
         bytesWritten, byteBuffer.limit());
     return bytesWritten;
   }
+
+  @Override
+  public int write(ByteBuffer src) throws IOException {
```

Comment:
How do we guarantee that all bytes requested are actually written?

Reply:
This should be fine. the write() interface is coming from WritableByteChannel and the caller will be handling this.